File: hackerrank/array/easy/java_solutions/CutTheSticks.java
package hackerrank.array.easy.java_solutions;

import java.util.LinkedList;
import java.util.List;

public class CutTheSticks {

    // https://www.hackerrank.com/challenges/cut-the-sticks/problem

    static class Solution {
        public static List<Integer> cutTheSticks(List<Integer> arr) {
            // Write your code here
            List<Integer> collector = new LinkedList<>();

            while (!arr.isEmpty()) {
                int n = arr.size();
                int min = 1 << 30;

                collector.add(n);

                for (Integer v : arr) {
                    min = Math.min(min, v);
                }
                List<Integer> newArr = new LinkedList<>();

                for (Integer v : arr) {
                    if (min < v) {
                        newArr.add(v - min);
                    }
                }
                arr = newArr;
            }
            return collector;
        }
    }
}

File: hackerrank/array/easy/golang_solutions/cut_the_sticks.go
package golang_solutions

// https://www.hackerrank.com/challenges/cut-the-sticks/problem

func cutTheSticks(arr []int32) []int32 {
	collector := make([]int32, 0)
	for len(arr) > 0 {
		min := int32(1 << 30)
		collector = append(collector, int32(len(arr)))
		newArr := make([]int32, 0)
		for _, v := range arr {
			min = minVal(min, v)
		}
		for _, v := range arr {
			if min < v {
				newArr = append(newArr, v-min)
			}
		}
		arr = newArr
	}
	return collector
}

func minVal(a, b int32) int32 {
	if a > b {
		return b
	}
	return a
}
